### PR TITLE
[ZEPPELIN-3582] Add type data to result of query from SQL

### PR DESCRIPTION
**Pull request by oxygen311**
### What is this PR for?
JDBCInterpreter knows type information for every SQL Query. We could save this info to pool and use it.
There are three types of table column:
- Number;
- String;
- Date.

### What type of PR is it?
Improvement

### What is the Jira issue?
[Zeppelin 3582](https://issues.apache.org/jira/projects/ZEPPELIN/issues/ZEPPELIN-3582)

### Screenshots
screenshot from 2018-07-06 18-20-29

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java [hunk 1]
```diff
@@ -115,10 +120,6 @@
   static final String STATEMENT_PRECODE_KEY_TEMPLATE = "%s.statementPrecode";
   static final String DOT = ".";
 
-  private static final char WHITESPACE = ' ';
-  private static final char NEWLINE = '\n';
-  private static final char TAB = '\t';
-  private static final String TABLE_MAGIC_TAG = "%table ";
```
Comment (felixcheung) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: why is this not needed?

Comment (oxygen311) [reply jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: It has been moved to JDBCUtils

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResultMessage.java [hunk 2]
```diff
@@ -16,18 +16,30 @@
...
  */
 public class InterpreterResultMessage implements Serializable {
   InterpreterResult.Type type;
   String data;
+  MessageColumnTypes msgColumnTypes;
 
   public InterpreterResultMessage(InterpreterResult.Type type, String data) {
     this.type = type;
     this.data = data;
+    this.msgColumnTypes = new MessageColumnTypes();
```
Comment (jongyoul) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResultMessage.java]: I think it's better to replace `this(type, data, new ArrayList<ColumnDef.TYPE>()`.

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResultMessage.java [hunk 3]
```diff
@@ -38,6 +50,14 @@ public String getData() {
     return data;
   }
 
+  public List<ColumnDef.TYPE> getListOfColumnTypes() {
+    return msgColumnTypes.getListOfColumnTypes();
+  }
+
+  public MessageColumnTypes getMessageColumnTypes() {
```
Comment (jongyoul) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResultMessage.java]: This looks used only one place inside `RemoteInterpreterServer`. Can we keep only `List<ColumnDef.TYPE>` as a member variable and make `MessageColumnTypes` in RemoteInterpreterServer` instead?

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResultMessage.java [hunk 4]
```diff
@@ -16,18 +16,30 @@
...
     this.type = type;
     this.data = data;
+    this.msgColumnTypes = new MessageColumnTypes();
+  }
+
+  public InterpreterResultMessage(InterpreterResult.Type type, String data,
+                                  List<ColumnDef.TYPE> columnTypes) {
+    this.type = type;
+    this.data = data;
+    this.msgColumnTypes = new MessageColumnTypes(columnTypes);
```
Comment (jongyoul) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResultMessage.java]: The below comments are related to this field. I think it's better to keep columnTypes itself.

**Comment by jongyoul** [comment]
Great!! Thank you!

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java (hunk 1 above)
Comment (felixcheung) [reply jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: I don't see the `TABLE_MAGIC_TAG = "%table ";`?

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResultMessage.java (hunk 3 above)
Comment (oxygen311) [reply zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResultMessage.java]: I think it's better to keep `MessageColumnTypes` class for better flexibility (e.g. custom `toString` method).

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResultMessage.java (hunk 2 above)
Comment (oxygen311) [reply zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResultMessage.java]: Done

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java (hunk 1 above)
Comment (oxygen311) [reply jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: It has been used only for knowing type in writer

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResultMessage.java (hunk 3 above)
Comment (jongyoul) [reply zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResultMessage.java]: Ah... What I meant is not to remove `MessageColumnTypes`. You don’t have to store it in this class by making new class. How about storing List... and making `MessageColumnTypes` if it’s needed.

Comment (oxygen311) [reply zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResultMessage.java]: Oh, ok, thx. Done

**Comment by jongyoul** [comment]
Thanks. Can you restart the failed one: 
https://travis-ci.org/TinkoffCreditSystems/zeppelin/builds/401673653

**Comment by zjffdu** [comment]
This is nice feature, could we also pass this kind of meta info to frontend ? so that frontend can display the data more intelligently.

**Comment by oxygen311** [comment]
@zjffdu we plan to do it in next pool requests.

File: jdbc/src/test/java/org/apache/zeppelin/jdbc/JDBCInterpreterTest.java [hunk 5]
```diff
@@ -658,4 +633,46 @@ public void testSplitSqlQueryWithComments() throws SQLException, IOException {
...
+    JDBCInterpreter t = new JDBCInterpreter(properties);
+    t.open();
+
+    String sqlQuery = "select * from test_table";
+
+    InterpreterResult interpreterResult = t.interpret(sqlQuery, interpreterContext);
+
+    assertEquals(InterpreterResult.Code.SUCCESS, interpreterResult.code());
+    assertEquals(Collections.nCopies(2, ColumnDef.TYPE.STRING),
+        interpreterResult.message().get(0).getColumnTypes());
```
Comment (zjffdu) [inline jdbc/src/test/java/org/apache/zeppelin/jdbc/JDBCInterpreterTest.java]: interpreter is not closed, it it better to put it in finally block

File: jdbc/src/test/java/org/apache/zeppelin/jdbc/JDBCInterpreterTest.java [hunk 6]
```diff
@@ -658,4 +633,46 @@ public void testSplitSqlQueryWithComments() throws SQLException, IOException {
...
+    JDBCInterpreter t = new JDBCInterpreter(properties);
+    t.open();
+
+    String sqlQuery = "select * from test_table_types";
+
+    InterpreterResult interpreterResult = t.interpret(sqlQuery, interpreterContext);
+
+    assertEquals(InterpreterResult.Code.SUCCESS, interpreterResult.code());
+    assertEquals(Arrays.asList(ColumnDef.TYPE.NUMBER, ColumnDef.TYPE.STRING, ColumnDef.TYPE.DATE),
+        interpreterResult.message().get(0).getColumnTypes());
```
Comment (zjffdu) [inline jdbc/src/test/java/org/apache/zeppelin/jdbc/JDBCInterpreterTest.java]: ditto.

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResultMessage.java [hunk 7]
```diff
@@ -16,18 +16,29 @@
...
+import java.util.ArrayList;
+import java.util.List;
 
 /**
  * Interpreter result message
  */
 public class InterpreterResultMessage implements Serializable {
   InterpreterResult.Type type;
   String data;
+  List<ColumnDef.TYPE> columnTypes;
```
Comment (zjffdu) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResultMessage.java]: Is it compatible with the previous InterpreterResult which is save into note.json ? I guess so, but it is better to confirm it.

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java [hunk 8]
```diff
@@ -649,13 +650,21 @@ public Object jobRun() throws Throwable {
...
                 context.getParagraphId(),
                 WellKnownResourceName.ZeppelinTableResult.toString(),
-                resultMessages.get(lastMessageIndex));
+                lastMessage
+            );
+            context.getResourcePool().put(
+                context.getNoteId(),
+                context.getParagraphId(),
+                WellKnownResourceName.ZeppelinTableType.toString(),
+                new MessageColumnTypes(lastMessage.getColumnTypes())
```
Comment (zjffdu) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java]: Shouldn't the lastMessage already include the columns types ? It is not necessary to put the columns types into ResourcePool again.

Comment (oxygen311) [reply zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java]: In that case we can't see type information in appropriate way:
screenshot from 2018-07-12 12-11-27

File: jdbc/src/test/java/org/apache/zeppelin/jdbc/JDBCInterpreterTest.java (hunk 5 above)
Comment (oxygen311) [reply jdbc/src/test/java/org/apache/zeppelin/jdbc/JDBCInterpreterTest.java]: Done

**Comment by jongyoul** [comment]
It seems to be ready to merge.

**Comment by zjffdu** [comment]
Sorry, I need go through it again today.

**Comment by jongyoul** [comment]
Got it. I wonder the status of this PR

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java [hunk 9]
```diff
@@ -725,9 +611,17 @@ private InterpreterResult executeSql(String propertyKey, String sql,
...
             } else {
               MutableBoolean isComplete = new MutableBoolean(true);
-              String results = getResults(resultSet,
-                  !containsIgnoreCase(sqlToExecute, EXPLAIN_PREDICATE), isComplete);
-              interpreterResult.add(results);
+
+              String results = getResults(resultSet, isComplete, getMaxResult());
+              List<ColumnDef.TYPE> columnTypes = getResultColumnTypes(resultSet);
+
+              final Boolean isTable = !containsIgnoreCase(sqlToExecute, EXPLAIN_PREDICATE);
```
Comment (zjffdu) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: It is fragile to use `containsIgnoreCase` to check whether the result is table.  The following code can break it.

e.g.
```
select * from table where a = '  Explain '
```

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java (hunk 8 above)
Comment (zjffdu) [reply zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java]: I am not sure what do you mean `nice type info`. It is issue of py4j. That we can only get JavaObject via py4j.

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java (hunk 9 above)
Comment (oxygen311) [reply jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: This remark does not apply to my changes. This line was already in zeppelin.

**Pull request closed without merging**